Add test for fast confirmation stability

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Adds a benchmarking setup to test load and confirmation performance over
time.
Adds a test to check that the stability and speed of the confirmations do not vary significantly under heavy load.

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

### Key places to review:
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->
- espresso/environment/benchmark/benchmark_helpers.go

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209731736667116

Ayiga committed:
Add test for fast confirmation stability

This commit adds a test to check the stability and speed provided by the
caff node and the l1 derived verifier, and checks to ensure that they don't
vary too greatly depending on the load.

dailinsubjam commented on espresso/environment/1_espresso_benchmark_test.go:
```diff
@@ -0,0 +1,114 @@
+		benchmark.WithCaffClient(system.NodeClient(env.RoleCaffNode)),
+		benchmark.WithVerifyClient(system.NodeClient(e2esys.RoleVerif)),
+		benchmark.AddSubmitter(benchmark.BenchmarkSubmitterConfig{
+			Interval: 100 * time.Millisecond,
```
Would be great if this is configurable, so that we can easily test out light load or heavier load case in the future.

Ayiga replied:
It is configurable by changing the interval for the "Submitters".  The lower the interval, the more transactions will be submitted.

dailinsubjam replied:
Oh right.

dailinsubjam commented on espresso/environment/1_espresso_benchmark_test.go:
```diff
@@ -0,0 +1,114 @@
+// TestE2eDevNetWithEspressoFastConfirmationStability is a test that attempts
+// to verify that the Espresso pipeline is able to perform consistently
+// under load without causing any significant delays in the confirmation
+// / finalization process.
```
Maybe a bit outside the scope of this PR, but it would be great to see a heavier load tested, since the task description mentions 'Run the rollup with maximum load.'

Ayiga replied:
It is unclear what "Maximum load" means.  But I'm open to suggestions.

philippecamacho replied:
Good suggestion Sishan, let us work on this aspect of the benchmarks once we have our devnet working.

dailinsubjam commented on espresso/environment/benchmark/benchmark_helpers.go:
```diff
@@ -0,0 +1,783 @@
+			}
+		}
+
+		receipt, err := wait.ForReceiptOK(ctx, client, txHash.Value)
```
I'm thinking whether there's any subtle difference between:
- The time of getting the tx receipt
- Finalization time of the tx
I think the former one means the client has already seen the tx is executed and included in a block, so this should be good to be used as the confirmation timestamp.

dailinsubjam commented on espresso/environment/benchmark/benchmark_helpers.go:
```diff
@@ -0,0 +1,783 @@
+		case receivedChain <- WithTimestamp(AnnotatedBlockHash{
+			BlockHash: header.Hash(),
+			Label:     label,
+		}):
```
Looks like we get timestamp from both transaction receipt and block header, what's the difference on the usage for these two and why not only use one?

Ayiga replied:
Ah.  The timestamp from the headers for the Caff Node and the L2 Verifier are used as an indicator as to when the block containing the transaction we previously received a receipt for has "appeared" on the respective client.

The idea here is that these transactions will appear on the L2 Sequencer itself roughly within the L2's Block Time.  However, the Caff Node and the L2 Verifier have timing factors that are independent of each other, and of the L2 Sequencer.

The limiting factor, in terms of time, for the Caff Node is entirely dependent on the Espresso processing time, and the time it takes for the Streamer to retrieve the blocks.

The limiting factor, in terms of time, for the L2 Verifier is entirely dependt on the L1 Block time.

So by collecting these measurements separately, we can readily compare them, and how the Espresso solution provides faster confirmation / derivation within the same data.

dailinsubjam replied:
~~Makes sense! Are they also collected by submitters? I'm just trying to find where we "by collecting these measurements separately, we can readily compare them"~~

dailinsubjam replied:
I got it now.

Ayiga committed:
Cleanup Benchmarker code

The benchmarker defines quite a lot of things that are not all grouped very
well.  These concepts should be split up into sensible groups.

Add Comments

The comments on the Benchmarker functions and primitives are lacking,
and incomplete.  Comments should be added to annotate, and document
the specific implementation details, and their reason for being.

Modify test benchmark statistics and variation targets

The benchmark statistics being checked against in the test have caff
receipt durations, and l2 verifier receipt durations that are weighted by
the total number of transactions in their respective blocks.  Additionally
these statistics lack any ability to track a block that does not contain the
submitted transactions.  This has been addressed by changing the
statistical measure computation.

The variation targets for the StdDev check were just set to 10% of the
mean.  This is a somewhat difficult target to pin down.  For now they have
been specified to not exceed 2 seconds instead.

philippecamacho commented on espresso/environment/1_espresso_benchmark_test.go:
```diff
@@ -0,0 +1,153 @@
+			t.Errorf("expected a small amount of variance in the submitted to receipt time:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+		}
+
+		if have, want := metrics.ReceiptToCaff.StdDev, 2*time.Second; have > want {
```
Should not it take more time to get the receipt from the Caff node, like around 4s?

Ayiga replied:
I do think that it might take longer than two seconds to get a verification from the Caff Node for a response.  However, in this case, we're not actually comparing the time it took for the Caff Node to provide a response.

Instead, we're looking at how much the time taken to get that response varies from the mean time to get the response.  The lower than number, the lower the variance and the higher the consistency.

philippecamacho replied:
Oh I missed `StdDev`. Thanks!

philippecamacho commented on espresso/environment/1_espresso_benchmark_test.go:
```diff
@@ -0,0 +1,153 @@
+			t.Errorf("expected a small amount of variance in the receipt to caff time:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+		}
+
+		if have, want := metrics.ReceiptToVerify.StdDev, 2*time.Second; have > want {
```
Should not it take more time to get the receipt from the OP node, like around 30s?

philippecamacho replied:
Answered [above](https://github.com/EspressoSystems/optimism-espresso-integration/pull/141#discussion_r2104656037).

philippecamacho reviewed (approved):
LGTM
